### PR TITLE
Refactor `prefer-array-index-of` and `no-array-for-each`

### PR DESCRIPTION
**Pull request by fisker**
Simplify `prefer-array-index-of` with an `isFunctionSelfUsedInside` function extracted from `no-array-for-each` rule.

### CODE REVIEW

File: rules/utils/is-function-self-used-inside.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+const getReferences = (scope, nodeOrName) => {
+	const {references = []} = findVariable(scope, nodeOrName) || {};
+	return references;
+};
+
+/**
+Check if `this`, `arguments`, or `id` of function is used inside it self.
+
+@param {Node} node - The function node.
```
Comment (sindresorhus) [inline rules/utils/is-function-self-used-inside.js:12]: These should match the parameter names in JS.

File: rules/utils/is-function-self-used-inside.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@param {Scope} [scope] - The scope of the function node.
```
Comment (sindresorhus) [inline rules/utils/is-function-self-used-inside.js:13]: This is marked as optional, but doesn't look like it is.

File: rules/utils/is-function-self-used-inside.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+const {findVariable} = require('eslint-utils');
+
+const getReferences = (scope, nodeOrName) => {
+	const {references = []} = findVariable(scope, nodeOrName) || {};
+	return references;
+};
+
+/**
+Check if `this`, `arguments`, or `id` of function is used inside it self.
```
Comment (sindresorhus) [inline rules/utils/is-function-self-used-inside.js:10]:
```suggestion
Check if `this`, `arguments`, or `id` of the function is used inside of itself.
```

Comment (sindresorhus) [reply rules/utils/is-function-self-used-inside.js:10]: Not clear from reading here what the "id" of the function is.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: rules/utils/is-function-self-used-inside.js (hunk 2 above)
Comment (fisker) [reply rules/utils/is-function-self-used-inside.js:13]: Right, I'll fix.

File: rules/utils/is-function-self-used-inside.js (hunk 3 above)
Comment (fisker) [reply rules/utils/is-function-self-used-inside.js]: Changed to ``Check if `this`, `arguments`, or the function name is used inside of itself.``

File: rules/utils/is-function-self-used-inside.js (hunk 2 above)
Comment (fisker) [reply rules/utils/is-function-self-used-inside.js]: Another thing in this function, I ask to pass two arguments, but acutally `functionNode` is alway `functionScope.block`, I just don't want people get the wrong scope. 

The scope can also get by calling `scopeManager.acquire(node)`, but in rule if it's in function, `context.getScope()` is easier to get the function scope, so I was considering accept `scope/scopeManager` as the second parameter.

Any better idea?

Comment (sindresorhus) [reply rules/utils/is-function-self-used-inside.js]: I think what you already have is the best solution.

**Pull request merged**